numpyro/infer/hmc_gibbs.py

```diff
@@ -156,42 +156,93 @@ def potential_fn(z_gibbs, z_hmc):
         return HMCGibbsState(z, hmc_state, rng_key)
 
 
-def _discrete_step(rng_key, z_discrete, pe, potential_fn, idx, support_size):
+def _discrete_gibbs_proposal_body_fn(z_init_flat, unravel_fn, pe_init, potential_fn, idx, i, val):
+    rng_key, z, pe, log_weight_sum = val
+    rng_key, rng_transition = random.split(rng_key)
+    proposal = jnp.where(i >= z_init_flat[idx], i + 1, i)
+    z_new_flat = ops.index_update(z_init_flat, idx, proposal)
+    z_new = unravel_fn(z_new_flat)
+    pe_new = potential_fn(z_new)
+    log_weight_new = pe_init - pe_new
+    # Handles the NaN case...
+    log_weight_new = jnp.where(jnp.isfinite(log_weight_new), log_weight_new, -jnp.inf)
+    # transition_prob = e^weight_new / (e^weight_logsumexp + e^weight_new)
+    transition_prob = expit(log_weight_new - log_weight_sum)
+    z, pe = cond(random.bernoulli(rng_transition, transition_prob),
+                 (z_new, pe_new), identity,
+                 (z, pe), identity)
+    log_weight_sum = jnp.logaddexp(log_weight_new, log_weight_sum)
+    return rng_key, z, pe, log_weight_sum
+
+
+def _discrete_gibbs_proposal(rng_key, z_discrete, pe, potential_fn, idx, support_size):
     # idx: current index of `z_discrete_flat` to update
     # support_size: support size of z_discrete at the index idx
 
-    # get z proposal
     z_discrete_flat, unravel_fn = ravel_pytree(z_discrete)
-
+    # Here we loop over the support of z_flat[idx] to get z_new
     # XXX: we can't vmap potential_fn over all proposals and sample from the conditional
     # categorical distribution because support_size is a traced value, i.e. its value
     # might change across different discrete variables;
     # so here we will loop over all proposals and use an online scheme to sample from
     # the conditional categorical distribution
-    def body_fn(i, val):
-        rng_key, z, pe, weight_logsumexp = val
-        rng_key, rng_accept = random.split(rng_key)
-        proposal = jnp.where(i >= z_discrete_flat[idx], i + 1, i)
-        z_new_flat = ops.index_update(z_discrete_flat, idx, proposal)
-        z_new = unravel_fn(z_new_flat)
-        pe_new = potential_fn(z_new)
-        weight_new = pe - pe_new
-        # Handles the NaN case...
-        weight_new = jnp.where(jnp.isfinite(weight_new), weight_new, -jnp.inf)
-        # transition_prob = e^weight_new / (e^weight_logsumexp + e^weight_new)
-        transition_prob = expit(weight_new - weight_logsumexp)
-        z, pe = cond(random.bernoulli(rng_accept, transition_prob),
-                     (z_new, pe_new), identity,
-                     (z, pe), identity)
-        weight_logsumexp = jnp.logaddexp(weight_new, weight_logsumexp)
-        return rng_key, z, pe, weight_logsumexp
-
+    body_fn = partial(_discrete_gibbs_proposal_body_fn,
+                      z_discrete_flat, unravel_fn, pe, potential_fn, idx)
     init_val = (rng_key, z_discrete, pe, jnp.array(0.))
```

Comment:
i'm a bit confused about why an initial log weight of `jnp.array(0.)` gives you the behavior you want

Reply:
This is Gibbs sampling, so we want to draw samples proportional to `e^-pe_i` (i.e. `log_weight_i = pe_init - pe_i` - the term `pe_init` is just a constant to scale down `pe_i` to improve precision) of each value `i` in the support. When `i = z_init`, `log_weight_i = pe_init - pe_init = 0.`.

Reply:
ok that makes sense. i guess it would ideally be `jnp.max(log_weights)` but that's unavailable because of the online scheme?

Reply:
Yes, you are right.